### PR TITLE
Added --follow argument in read command

### PR DESCRIPTION
**Pull request by approxit**
What I've done:
- Moved CLI stuff to dedicated module, as its a better place instead of keeping everything in `__main__.py`
- Simplified `read` to be a single command with `file-type` argument instead of be a command-group with duplicated command with only 1 string difference
- Added `read_file_iter` to manager and storage objects
- Tests to cover basic functionality

Additional stuff:
- While following a log, what to do in case of application dies?
  Currently we're ignoring app state and monitor file anyways
- While following a log, what to do in case of application is being purged?
  Currently file is continuously opened while following, hence preventing file deletion via purge

### CODE REVIEW

File: tests/test_manager_api.py [hunk 1]
```diff
@@ -111,6 +111,31 @@ def test_read_file(get_dapp):
...
+        iterator = dapp.read_file_iter(file_type, False)
+
+        try:
+            with open(mock_fname) as f:
+                assert next(iterator) == f.read()
+
+            with dapp.storage.file_name(file_type).open("a") as file:
+                file.write(f"some\nmore\n{file_type}\nlines!")
+
+            assert next(iterator) == f"some\nmore\n{file_type}\nlines!"
```
Comment (lucekdudek) [inline tests/test_manager_api.py:133]: File content can be moved to a separated variable `f"some\nmore\n{file_type}\nlines!"`

File: tests/test_manager_api.py [hunk 2]
```diff
@@ -111,6 +111,31 @@ def test_read_file(get_dapp):
             assert dapp.read_file(file_type) == f.read()
 
 
+@pytest.mark.parametrize("get_dapp", get_dapp_scenarios)
+def test_read_file_iter(get_dapp):
```
Comment (lucekdudek) [inline tests/test_manager_api.py:115]: Having a test that covers chunking would be nice.

File: tests/test_manager_api.py [hunk 3]
```diff
@@ -111,6 +111,31 @@ def test_read_file(get_dapp):
...
+    dapp = get_dapp(
+        [sys.executable, asset_path("write_mock_files.py")], state_file=True, data_file=True
+    )
+    sleep(0.5)
+
+    for file_type in ("state", "data"):
```
Comment (lucekdudek) [inline tests/test_manager_api.py:121]: It looks like two separated unit tests. Parameterizing can improve readability of this test

File: pyproject.toml [hunk 4]
```diff
@@ -33,7 +33,7 @@ python = "^3.8"
 
 psutil = "^5.9"
 appdirs = "^1.4"
-click = "^8.1"
+click = "^7.0"
```
Comment (lucekdudek) [inline pyproject.toml:36]: We usually add comments in `toml` on why we keep some packages back.

File: dapp_manager/cli.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
+import sys
```
Comment (shadeofblue) [inline dapp_manager/cli.py:1]: I don't feel any strong benefit from moving the commands to a separate module... especially given that CLI _is_ the only interface that the `dapp-manager` offers... I won't protest though...

**Comment by shadeofblue** [comment]
re: questions

>     While following a log, what to do in case of application dies?
>     Currently we're ignoring app state and monitor file anyways

hmm, app state is irrelevant imvho for this scenario but indeed, dapp-runner's death probably should be detected and following stopped...
anyway, it wouldn't matter much on its own but the second scenario is more important... 

>     While following a log, what to do in case of application is being purged?
>     Currently file is continuously opened while following, hence preventing file deletion via purge

I would think this to be a more important issue... and detecting the death of `dapp-runner`s process should resolve it

**Review by shadeofblue — APPROVED** [review]
generally, looks good, though, as mentioned before, I don't feel it was necessary to move the CLI commands to their own module...

File: tests/test_manager_api.py [hunk 6]
```diff
@@ -99,16 +99,76 @@ def test_stop_timeout_kill(get_dapp):
...
+    additional_line = f"some\nmore\n{file_type}\nlines!"
+
+    iterator = dapp.read_file_iter(file_type, False)
+
+    try:
+        assert next(iterator) == "THIS "
+        assert next(iterator) == "IS\nA\n"
+        assert next(iterator) == "STATE"
+        assert next(iterator) == " FILE"
+        assert next(iterator) == "!\n"
```
Comment (lucekdudek) [inline tests/test_manager_api.py:159]: I have no idea from where is this `"THIS ""IS\nA\n""STATE"" FILE""!\n"`

**Comment by approxit** [comment]
Refactor time!

Current state for this PR is in 2 commits. Unfortunately GH can't display diff properly (IDE works better for this), so to lool up changes in `__main__.py`/`cli.py` view only the first commit.

What I've done:
- As behavior for found edge cases was decided, I took the change to clean code more significantly than touch only smallest changes.
- Implemented chunking in stream reader functions
- Reading from streams are now iterators (due to chunking)
- Reading from streams are now using same lower level function from storage
- `read_file_follow` have additional behavior to match detected edge cases
- `Storage.iter_file_chunks` will close stream file before returning trailing chunk of data. This helps with handling file following,
- Tests to have full coverage of read functions
- Minor function renamings to have consistent naming

File: dapp_manager/cli.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+@_capture_api_exceptions
+def read(app_id: str, file_type: RunnerReadFileType, ensure_alive: bool, follow: bool):
+    """Read output from the given app."""
+
+    dapp = DappManager(app_id)
+
+    read_function: Callable = dapp.read_file_follow if follow else dapp.read_file
+
+    for chunk in read_function(file_type, ensure_alive=ensure_alive):
+        print(chunk, end="")
```
Comment (lucekdudek) [inline dapp_manager/cli.py:164]: I am not a fan of new `read_file` interface. I would keep the `read_file` as a return function. `sample_usage.py` showcases that it got more complicated and I cannot find a reason why.

Here we could simply implement it like this 
```python
if follow is True:
    for chunk in read_function(file_type, ensure_alive=ensure_alive):
        print(chunk, end="")
else:
    print(dapp.read_file)
```

File: dapp_manager/dapp_starter.py [hunk 8]
```diff
@@ -36,8 +36,15 @@ def start(self, timeout: float) -> None:
...
         if not success:
-            runner_stdout = self.storage.read_file("stdout")
-            runner_stderr = self.storage.read_file("stderr")
+            try:
+                runner_stdout = "".join(data for _, data in self.storage.iter_file_chunks("stdout"))
+            except FileNotFoundError:
+                runner_stdout = None
+
+            try:
+                runner_stderr = "".join(data for _, data in self.storage.iter_file_chunks("stderr"))
```
Comment (lucekdudek) [inline dapp_manager/dapp_starter.py:45]: Raw string manipulation here bothers me. Having additional method in `storage` that return complete content of a file would work better IMO. `self.storage.read_file("stdout")` was more cleaner.

**Review by lucekdudek — COMMENTED** [review]
I find new interfaces of simple read file methods more complicated then the old ones. Both in storage and dapp_runner. I would prefer to keep interaction between modules as simple a possible.

File: dapp_manager/dapp_manager.py [hunk 9]
```diff
@@ -7,18 +7,20 @@
...
 
 from .dapp_starter import DappStarter
 from .exceptions import AppNotRunning
-from .storage import RunnerFileType, SimpleStorage
+from .storage import RunnerReadFileType, SimpleStorage
 
 PathType = Union[str, os.PathLike]
 
-COMMAND_OUTPUT_INTERVAL = 1.0
+COMMAND_OUTPUT_INTERVAL = timedelta(seconds=1)
```
Comment (shadeofblue) [inline dapp_manager/dapp_manager.py:21]: hmm... again, I see no strong benefit from forcing this value to be of `timedelta` type... to me it's obvious that - if it's a scalar float, it's the number of seconds... and we never use it in a way that would make us benefit from it being a `timedelta` (that would be the case if we wanted to perform comparisons between `datetime`s, etc)

File: dapp_manager/exceptions.py [hunk 10]
```diff
@@ -42,17 +45,27 @@ class StartupFailed(DappManagerException):
...
-            f"{runner_stderr}\n\n"
-            "--- dapp-runner stdout ---\n"
-            f"{runner_stdout}\n\n"
-            "--- pre-runner stderr ---\n"
-            f"{stderr}\n\n"
-            "--- pre-runner stdout ---\n"
-            f"{runner_stdout}\n\n"
+    def __init__(self, stdout, stderr, runner_stdout: Optional[str], runner_stderr: Optional[str]):
+        if runner_stderr is None:
+            runner_stderr_text = "--- no dapp-runner stderr ---"
```
Comment (shadeofblue) [inline dapp_manager/exceptions.py:50]: in such case, I'd just have empty string here... so that we don't clutter the output... 

```suggestion
            runner_stderr_text = ""
```

File: dapp_manager/exceptions.py [hunk 11]
```diff
@@ -42,17 +45,27 @@ class StartupFailed(DappManagerException):
...
+        else:
+            runner_stderr_text = f"--- dapp-runner stderr ---\n{runner_stderr}"
+
+        if runner_stdout is None:
+            runner_stdout_text = "--- no dapp-runner stdout ---"
```
Comment (shadeofblue) [inline dapp_manager/exceptions.py:55]: same here... 

```suggestion
            runner_stdout_text = ""
```

**Comment by shadeofblue** [comment]
added a couple of comments, otherwise, I somewhat share @lucekdudek 's sentiment with regards to the interface changes... though, given that we now have two ways to read file's contents, maybe it's inevitable...

File: dapp_manager/dapp_manager.py (hunk 9 above)
Comment (approxit) [reply dapp_manager/dapp_manager.py:21]: As discussed offline, using `timedelta` looks more friendly than sole value.

**Review by lucekdudek — APPROVED** [review]
